### ISSUE TITLE
Address issues w/ shutting down a context used across threads

### ISSUE THREAD

**Issue by daveyarwood** [daveyarwood/ezzmq#1]
See: https://github.com/zeromq/jeromq/issues/116, https://github.com/zeromq/jeromq/issues/147

All blocking calls (sending, receiving, polling) apparently need to be wrapped in a try/catch for ZMQException and if the error code is ETERM, then the following needs to happen:
- All sockets must be setLinger'd to 0 and closed.
  - ...but calling setLinger in this scenario throws a CtxTerminatedException, so you have to catch _that_ and ignore it.
- The context needs to be terminated.

> Note: ZContext does the setLinger-and-catch-CtxTerminatedException-and-ignore-it thing for you when it is destroyed, so I think for ZContext all you have to do is terminate the context.

It's annoying that we have to do all of this manually. I'm thinking this boilerplate can be abstracted away, and the best place to do it is probably in the `with-context` / `with-new-context` macros. Can probably just wrap all the generated code in a big ol' try-catch and do the above whenever a ZMQException is caught and the error code is ETERM.

Should add a test case to confirm whether this works properly. I believe the use case has to do with using a single ZMQ context and multiple inproc sockets running on different threads, then terminating the context on the main thread and ensuring that the other threads shut down cleanly.

**Comment by daveyarwood** [comment]
Experimenting with this by porting the [zguide interrupt example](https://github.com/zeromq/jeromq/blob/master/src/test/java/guide/interrupt.java), here are some findings:
- The idiomatic way to create a worker thread in Clojure is with a `future`. Interestingly, a Clojure program seems to just exit before a future is complete, unless you dereference it, in which case the program will block until the future returns a value.
- You can also call `future-cancel`, which throws a CancellationException if the future isn't done yet. It doesn't seem to be possible catch this exception from the main thread.
- We also can't catch the ETERM in the main thread because it's happening in the worker thread.
- Adding a try/catch, just like in the zguide interrupt example, works. You can catch ZMQException, check if its error code is the same as ZMQ.Error.ETERM, and if it is, exit the work loop.

So it seems like the special behavior is something you need to add inside a worker thread when you need one.

Maybe the thing to do is to provide some kind of `worker-thread` macro, which allows you to provide an (optional) shutdown function, and abstract away the try/catch code so that your shutdown function (if there is one) gets called if ETERM is thrown at any point in that thread.

Something like:

``` clojure
(worker-thread {:on-interrupt #(println "Interrupted!")}
  (do-work-here))
```

**Comment by daveyarwood** [comment]
`worker-thread` macro available in 0.2.0.

[Translation of the zguide interrupt example](https://github.com/daveyarwood/ezzmq/blob/master/examples/zguide/interrupt.boot) illustrates its use.
